crystal-lang/crystal
# Don't use integer unions in the compiler when analyzing enum values

This is a workaround for #10359

It seems there's a bug that is triggered when unions of integer types are involved.

There are already many bugs in the compiler which will eventually be fixed. The problem with #10359 is that the code that triggers the bug is also present in the compiler. And because of that, compiling a compiler in release mode produces a compiler that can no longer compile any program that depends on "prelude".

We have two choices:

1. We fix the bug. That means the bug will no longer be present in the compiler, or in any program.
2. We avoid using the code that triggers the bug in the compiler. That means that programs that include this code pattern might not behave well, but the compiler can continue to be compiled.

I think we can do point 2 to unblock the compiler, and to unblock the 1.0 release. Then we can do point 1 with time, similar to other bugs where we take our time to fix them in order to understand the problem.

There's another reason to prefer going with a workaround: I believe the bug is triggered when unions of multiple integer types, and `nil`, are involved. This should be pretty rare in most programs, because one doesn't need to work with multiple integer types (unlike the compiler). In fact the compiler could always do math using `Int64` and only go to perform math in other types to check for overflow errors, but then go back to `Int64`, maybe even disallowing `UInt64` values that don't fit into an `Int64`. But given that the language allows such union types, it was implemented like that.

This PR implements point 2: it avoids using integer union types when computing the value of enum members.

The workaround can be later undone once we actually fix the bug.

This PR also adds LLVM 11.1.0 support, which is just an entry in `llvm-versions.txt`. I needed that because I have LLVM 11.1.0 on my machine and without this change I couldn't work on the compiler.

# Steps for testing that this workaround is effective

1. Run `make clean crystal release=1`. Make sure you are using LLVM 11 or bigger.
2. Using `bin/crystal` compile the following program:

```crystal
enum Test : UInt32
  One   =  1
  Two   =  2
  Three =  3
  Four  =  4
  Five  =  5
  Six   =  6
  Seven =  7
  Eight =  8
  Nine  =  9
  Ten   = 10

  {% for member in @type.constants %}
    {% puts member, @type.constant(member) %}
  {% end %}
end
```

The output should be:

```
One
1_u32
Two
2_u32
Three
3_u32
Four
4_u32
Five
5_u32
Six
6_u32
Seven
7_u32
Eight
8_u32
Nine
9_u32
Ten
10_u32
```

Before this PR the output was:

```
One
1_u32
Two
8_u32
Three
9_u32
Four
0_u32
Five
1_u32
Six
2_u32
Seven
3_u32
Eight
4_u32
Nine
5_u32
Ten
4294967306_u32
```
